[#47] Change csr to property

sam-bertin opened the pull request:
> [!NOTE]  
> Closes #47 

I have based this work on the pull request #46.

🎨 This will help keep the signing request values in sync with the generated CSR.
🛂 Also restricts the permissions depending on the singing request status (signed / not signed). 
✅ Updated the tests to match the new implementation.

sam-bertin committed:
[#47] Changed CSR to proprety

This will help keep values in sync.
Also restricts the permissions depending on the singing request status (signed / not signed).  Updated the test to match the new implementation.

CharString commented on simple_certmanager/models.py:
```diff
@@ -71,26 +65,28 @@ def __str__(self):
         # generate private key once
         if not self.private_key:
-            self.private_key, self.csr = generate_private_key_with_csr(
+            self.private_key, csr = generate_private_key_with_csr(
```
Or just

```suggestion
            self.private_key = generate_private_key(
```

and delete `generate_private_key_with_csr`, I don't think it's used anymore

CharString commented on tests/test_csr.py:
```diff
@@ -362,7 +360,7 @@ def test_csr_does_not_renew_if_subject_dont_change(admin_client):
     assert signing_request.private_key == original_pk
     # CSR should not have changed since suject and pk are the same
-    assert signing_request.csr != orginal_csr
+    assert signing_request.csr == orginal_csr
```
Hmmm weird. ow did this pass? The assertion contradicts the comment.

Maybe rename this test to `test_csr_does_not_change_if_subject_dont_change`. "Renewal" is a loaded term in this context. We don't want to confuse this with "certificate renewal".

sam-bertin replied:
frankly I had no idea why it passed before, I got pretty confused when reading the comment and confirming that this test was not showing the expected behavior.

sergei-maertens replied:
my curiosity would like to understand what's going on here since the whole csr-as-a-property hangs on this proof :stuck_out_tongue:

CharString commented on tests/test_csr.py:
```diff
@@ -451,3 +449,44 @@ def test_saving_public_certifate_disables_signing_request_fields(admin_client):
+    assert response.status_code == 200
+
+    signing_request = SigningRequest.objects.get()
+    assert signing_request is not None
```
The default `django.db.models.Manager.get()` raises an exception, it never returns None.

CharString commented on simple_certmanager/admin.py:
```diff
@@ -92,6 +94,51 @@ class SigningRequestAdmin(admin.ModelAdmin):
+        return my_urls + urls
+
+    def download_csr(self, request, pk):
+        # Here I don't use the change permission because it won't be given after
+        # the CSR is signed and the certificate is uploaded.
+        if self.has_add_permission(request):
```
```suggestion
        if self.has_view_permission(request):
```

It's what we're doing.

CharString commented on simple_certmanager/admin.py:
```diff
@@ -92,6 +94,51 @@ class SigningRequestAdmin(admin.ModelAdmin):
+    def csr_link(self, obj: SigningRequest):
+        # Here I check that the common name is not empty
+        # because the CSR can't be generated without it and will only be generated
+        # after the object is saved the first time.
```
Nice comment. Should we do a similar check in the download_csr view? Do I get a 404 now if there's no common_name? Is it even possible? `SigningRequest.common_name` doesn't have `blank=True`, so `""` is not a valid value.

sam-bertin replied:
if there is no common_name that means we are creating the instance. The download_csr is only used after the model has been saved at least once. A saved model will hold a csr.

You won't get 404 as the link is not visible as long as there is no common_name.

![image](https://github.com/user-attachments/assets/a1fcfff5-3a8b-468b-afed-36078e45ab40)

CharString commented on tests/test_csr.py:
```diff
@@ -451,3 +449,44 @@ def test_saving_public_certifate_disables_signing_request_fields(admin_client):
+    assert response.status_code == 302
+    base_url = "/admin/login/?next=/admin/simple_certmanager/"
+    url = base_url + "signingrequest/dowload_csr/{}".format(signing_request.pk)
+    assert response.url == url
```
I feel this test asserts a lot of unrelated. Can it be logically be split up in smaller tests? Or am I misreading and in this sequence a unit?

CharString reviewed (changes requested):
I like the comments in this one; they tell the "why". I can read code, so I never feel the need for "what", but "why" is what makes me hunt through git blame and tickets.

If you feel the need to comment "what", it's probably because the code is murky and could be reformulated.

sam-bertin committed:
👀 Processing review

Applying suggestions from Chris review:
- ✅ decoupled tests
- 🔥 removed generate_csr_with_private_key_method (unused)
- 🛂 permission check update

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -92,6 +94,51 @@ class SigningRequestAdmin(admin.ModelAdmin):
+        urls = super().get_urls()
+        my_urls = [
+            path(
+                "download_csr/<int:pk>",
```
```suggestion
                "download_csr/<int:pk>/",
```
purely cosmetic :) but we like our trailing slashes :D

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -92,6 +94,51 @@ class SigningRequestAdmin(admin.ModelAdmin):
+                    reverse("admin:download_csr", args=(obj.pk,))
+                )
+            )
+        return _("CSR not generated yet")
```
```suggestion
        return _("Save the signing request to be able to download it.")
```

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -92,6 +94,51 @@ class SigningRequestAdmin(admin.ModelAdmin):
+        # Here I check that the common name is not empty
+        # because the CSR can't be generated without it and will only be generated
+        # after the object is saved the first time.
+        if obj.common_name:
```
you could also check `obj.pk is not None` which is the canonical way to check if the object has been saved or not, but the semantic check for `common_name` is better!

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -92,6 +94,51 @@ class SigningRequestAdmin(admin.ModelAdmin):
+                '<a href="{}">Download CSR</a>'.format(
+                    reverse("admin:download_csr", args=(obj.pk,))
+                )
+            )
```
do not call `format` yourself when using `format_html`, that's approach is usually vulnerable to XSS. Double check the docs for the correct usage :wink: 

Additionally, make sure that the link content is translatable :)

sam-bertin committed:
👀 Processing review

Applying suggestions from Sergei
- 🔒️proper format_html usage
- 🚸 user translatable missing content
- 🎨 code styling
